### PR TITLE
ZMQ retries on EINTR errors

### PR DESCRIPTION
**Pull request by dgant**
* Added 10x (**EDIT:** Infinite) retries to EINTR errors, which seem to sometimes occur intermittently (non-fatally) and sometimes fatally
* Consolidated duplicate copies of zmq.hpp

### CODE REVIEW

File: client/connection.h [hunk 1]
```diff
@@ -11,7 +11,7 @@
 
 #include <mutex>
 
-#include "zmq.hpp"
+#include "../BWEnv/include/zmq.hpp"
```
Comment (edran) [inline client/connection.h]: Tbh I would have liked to do the opposite - i.e. for BWEnv to look for `torchcraft/include/zmq.hpp`, but this is outside of the scope of the PR, and it makes sense to standardise this way right now.

**Comment by jgehring** [comment]
Please don't modify `zmq.hpp`. `recv()` in there should be a simple wrapper around `zmq_recv()`. Throwing with `EINTR` is perfectly valid -- it's not mysterious either, since that's what happens if you get interrupted by a signal. Please check for EINTR in `Connection::receive()` and retry there as long as you keep receiving it.

**Comment by dgant** [comment]
@jgehring Should it be in zmq_server as well?

File: client/connection.h (hunk 1 above)
Comment (dgant) [reply client/connection.h]: Not a hard change to make, though zmq.hpp is in /client/ right now, not /include/.

Comment (edran) [reply client/connection.h]: Our include paths are all over the place right now, so I wouldn't do this in this PR.

**Comment by jgehring** [comment]
> Should zmq_server retry as well?

It would not hurt but then again there should be no immediate need for that. AFAIK OpenBW does not catch any signals, and I don't have any idea about signal handling on Windows.

File: client/connection.cpp [hunk 2]
```diff
@@ -92,41 +92,38 @@ bool Connection::send(const void* buf, size_t len) {
...
+    if (exception.num() == EINTR) {
+      return receive_(assignToDestination);
+    }
+    errnum_ = exception.num();
+    errmsg_ = exception.what();
     return false;
   }
 }
 
+bool Connection::receive(std::string& dest) {
```
Comment (jgehring) [inline client/connection.cpp]: I think this function is unused -- let's remove it?

File: client/connection.cpp [hunk 3]
```diff
@@ -92,41 +92,38 @@ bool Connection::send(const void* buf, size_t len) {
...
       errmsg_ = ERRMSG_TIMEOUT_EXCEEDED;
     }
-    return res;
-  } catch (zmq::error_t& e) {
-    errnum_ = e.num();
-    errmsg_ = e.what();
+    return response;
+  } catch (zmq::error_t& exception) {
+    if (exception.num() == EINTR) {
+      return receive_(assignToDestination);
```
Comment (jgehring) [inline client/connection.cpp]: I'd prefer a loop to be honest, you could just continue here since all other code paths will return.

Comment (dgant) [reply client/connection.cpp]: Why a loop? Modern C++ compilers have tail-call optimization.

File: client/connection.h [hunk 4]
```diff
@@ -83,6 +83,11 @@ class Connection {
   /// @param dest [out] Vector to store the received data to
   /// @return true if the receive operation succeeded
   bool receive(std::vector<uint8_t>& dest);
+  
+ private: 
+  bool receive_(std::function<void()> assignToDestination);
```
Comment (jgehring) [inline client/connection.h]: This can be removed now?
